PP-5984 Fix validation parity on user entered reference

sfount opened the pull request:
Current validation on payment link reference doesn't make sense, there
is both client and server side validation (with useful error messages)
for some validation -- there appears to be no client side validation for
length; the `max-length` HTML attribute is used which won't allow the
user to enter something too long but will also not give any meaningful
feedback (instead just truncating whatever is entered).

* Remove max length assertion - allow the server side
validation to provide the user with a meaningful message
* Reduce the max length to 50, the Products database stores this in
`reference_number` which is currently restricted to `VARCHAR(50)`
https://github.com/alphagov/pay-products/blob/master/src/main/resources/migrations/00029_alter_column_reference_number_length_payments.sql#L4

---

Note Connector allows references of up to 250 characters `VARCHAR(250)`
https://github.com/alphagov/pay-connector/blob/master/src/main/resources/migrations.xml#L68,
there doesn't seem to be a reason Connector should support this and
Products shouldn't. These database columns should and related
validations should be brought into parity with each other.

--- 

TL;DR client side validation for length doesn't exist, remove this. Products database
is currently configured to `VARCHAR(50)`, initially limit reference to 50 characters.

alexbishop1 commented on test/unit/controllers/payment_reference/post_product_reference_controller_test.js:
```diff
@@ -94,7 +94,7 @@ describe('product reference post controller', function () {
   })
 
-  describe('when reference field exceeds max length 255', function () {
+  describe('when reference field exceeds max length', function () {
```
Could you update the actual tested value to be just over the limit (i.e. 51 characters)?

sfount commented:
@alexbishop1 thanks, agreed! We'll need both this change and the db migration to get the validation user experience right. Addressed comments https://github.com/alphagov/pay-products-ui/pull/699/commits/1692021a1558cd269811e984af18a26c8311a3a9

alexbishop1 commented on test/unit/controllers/payment_reference/post_product_reference_controller_test.js:
```diff
@@ -105,7 +105,7 @@ describe('product reference post controller', function () {
       nock(config.ADMINUSERS_URL).get(`/v1/api/services?gatewayAccountId=${product.gateway_account_id}`).reply(200, service)
 
-      const referenceNumber = 'This_is_a_256_characters_long_String_This_is_a_256_characters_long_String_This_is_a_256_characters_long_String_This_is_a_256_characters_long_String_This_is_a_256_characters_long_String_This_is_a_256_characters_long_String_This_is_a_256_characters_long_Stri'
+      const referenceNumber = 'This_is_a_51_characters_long_String_This_is_a_51_ch'
```
Missed a trick by not using `'This string is exactly 51 characters long in length'`!
